### ISSUE TITLE
Running `bundle install --jobs 4 --retry 3` intermittently throws `[ruby] WARNING OutOfMemoryError`

### ISSUE THREAD

**Issue by MSP-Greg** [oracle/truffleruby#2127]
Using truffleruby-head with GitHub Actions, see:
https://github.com/MSP-Greg/puma/runs/1288146881?check_suite_focus=true#step:5:9

Maybe something with zlib?  Some of the log:
```
TypeError: class or module required for rescue clause
  /home/runner/.rubies/truffleruby-head/lib/truffle/truffle/cext.rb:1503:in `rb_rescue2'
  exception.c:90:in `rb_rescue2'
  zlib.c:2409:in `gzfile_read_raw'
  zlib.c:2694:in `gzfile_read_more'
  zlib.c:2825:in `gzfile_read_all'
  zlib.c:3828:in `rb_gzreader_read'
  /home/runner/.rubies/truffleruby-head/lib/truffle/truffle/cext_ruby.rb:39:in `read'
```

Reminder - this is an intermittent error...

**Comment by eregon** [comment]
The `rb_rescue2` TypeError is from
https://github.com/oracle/truffleruby/blob/d84da65fc8531aca9e1e397e1a75eee83834b655/src/main/c/zlib/zlib.c#L2409-L2411
and
https://github.com/oracle/truffleruby/blob/d84da65fc8531aca9e1e397e1a75eee83834b655/src/main/c/cext/exception.c#L79-L91
and
https://github.com/oracle/truffleruby/blob/d84da65fc8531aca9e1e397e1a75eee83834b655/lib/truffle/truffle/cext.rb#L1503

Probably we should filter out `(VALUE) 0` which is the same as `Qfalse` from `rescued`.

**Comment by MSP-Greg** [comment]
@eregon - I have no idea, just the messenger...  Thanks.

**Comment by eregon** [comment]
Yeah I know, just making notes of what I found and how to help fix this :)

**Comment by eregon** [comment]
The actual `rescue2` error is fixed in 953cc2173723074a388a01fe12cf8ea39094ed0f, thanks for the report.
(the spec for it unfortunately checked for a TypeError going through, and we raised another TypeError due to the bug, which got unnoticed)

I'd like to close this issue, but feel free to reopen or comment if an OutOfMemoryError happens again, the `rescue2` might not be the root cause of the OutOfMemoryError.

**Comment by MSP-Greg** [comment]
@eregon

Thanks.  Oddly, I just saw the error in [nio4r CI](https://github.com/socketry/nio4r/runs/1350376820#step:5:9) yesterday.  I knew you'd get to it sometime.  If I see it again, I'll reopen.

Question - would you suggest running CI with truffleruby-head as a normal job (not 'allow-failure')?

The question is somewhat clouded since 'allow-failure' doesn't really work in Actions...

**Comment by eregon** [comment]
I saw it in https://github.com/ruby/setup-ruby/runs/1373979628?check_suite_focus=true
 ```
/home/runner/.rubies/truffleruby-head/bin/bundle install --jobs 4
  [ruby] WARNING OutOfMemoryError
  NoMemoryError: failed to allocate memory
    <internal:core> core/string.rb:44:in `byteslice'
    /home/runner/.rubies/truffleruby-head/lib/truffle/stringio.rb:421:in `read'
    /home/runner/.rubies/truffleruby-head/lib/truffle/stringio.rb:533:in `sysread'
    /home/runner/.rubies/truffleruby-head/lib/truffle/truffle/cext.rb:847:in `rb_funcall'
    /home/runner/.rubies/truffleruby-head/lib/truffle/truffle/cext.rb:840:in `rb_funcallv'
    call.c:18:in `rb_funcallv'
    zlib.c:2380:in `gzfile_read_raw_partial'
    /home/runner/.rubies/truffleruby-head/lib/truffle/truffle/cext.rb:1484:in `rb_rescue2'
    exception.c:94:in `rb_rescue2'
    zlib.c:2409:in `gzfile_read_raw'
    zlib.c:2694:in `gzfile_read_more'
    zlib.c:2825:in `gzfile_read_all'
    zlib.c:3828:in `rb_gzreader_read'
    /home/runner/.rubies/truffleruby-head/lib/truffle/truffle/cext_ruby.rb:39:in `read'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/compact_index_client/updater.rb:58:in `block in update'
    /home/runner/.rubies/truffleruby-head/lib/mri/tmpdir.rb:93:in `mktmpdir'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/compact_index_client/updater.rb:31:in `update'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/compact_index_client.rb:98:in `update'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/compact_index_client.rb:85:in `update_and_parse_checksums!'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/fetcher/compact_index.rb:75:in `available?'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/fetcher/compact_index.rb:16:in `call'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/fetcher/compact_index.rb:16:in `available?'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/fetcher.rb:162:in `use_api'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/source/rubygems.rb:388:in `block in api_fetchers'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/source/rubygems.rb:388:in `select'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/source/rubygems.rb:388:in `api_fetchers'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/source/rubygems.rb:393:in `block in remote_specs'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/index.rb:11:in `build'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/source/rubygems.rb:392:in `remote_specs'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/source/rubygems.rb:88:in `specs'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/definition.rb:273:in `block (2 levels) in index'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/definition.rb:271:in `each'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/definition.rb:271:in `block in index'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/index.rb:11:in `build'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/definition.rb:268:in `index'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/definition.rb:258:in `resolve'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/definition.rb:170:in `specs'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/definition.rb:158:in `resolve_remotely!'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/installer.rb:303:in `resolve_if_needed'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/installer.rb:84:in `block in run'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/process_lock.rb:12:in `block in lock'
    <internal:core> core/io.rb:688:in `open'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/process_lock.rb:9:in `lock'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/installer.rb:73:in `run'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/installer.rb:25:in `install'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/cli/install.rb:66:in `run'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:256:in `block in install'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/settings.rb:124:in `temporary'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:255:in `install'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:30:in `dispatch'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:24:in `start'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/exe/bundle:46:in `block in <top (required)>'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'
    /home/runner/.rubies/truffleruby-head/lib/gems/gems/bundler-2.1.4/exe/bundle:34:in `<top (required)>'
    <internal:core> core/kernel.rb:401:in `load'
    <internal:core> core/kernel.rb:401:in `load'
    /home/runner/.rubies/truffleruby-head/bin/bundle:23:in `<main>'
 ```

**Comment by eregon** [comment]
It was quite tricky to reproduce, because this OOM actually happens when downloading the full index with all gem versions, and that gets cached. But on GitHub Actions, that index is not cached, if there is no Gemfile.lock, and so in that case is downloading once per workflow job.
That index is 5957796 bytes (5.7MB) compressed, 15513690 bytes (14.8MB) uncompressed, and it's read fully in the Ruby heap here:
https://github.com/oracle/truffleruby/blob/69dc4f183d70fb40283fa6020c81b892a28b9f3f/lib/mri/bundler/compact_index_client/updater.rb#L56-L58

So something in that seems particularly inefficient in memory footprint.

This reproduces it reliably:
```ruby
# Gemfile
source 'https://rubygems.org'

gem "rake"
gem "path"
gem "json", "2.2.0"
```

```ruby
# once is enough
bundle config --local path vendor/bundle

rm -rf Gemfile.lock vendor/bundle ~/.bundle
TRUFFLERUBYOPT="--vm.Xmx1g" DEBUG=true bundle install --jobs 4
```

With `--vm.Xmx2g` it seems to work fine
`DEBUG=true` is optional, it just enables debug messages from Bundler, showing what it downloads.

**Comment by eregon** [comment]
Simpler reproduction:
```ruby
require 'net/http'

# https://index.rubygems.org/versions
Net::HTTP.start('index.rubygems.org', 443, use_ssl: true) do |http|
  headers = {"Accept-Encoding"=>"gzip"}
  request = Net::HTTP::Get.new "/versions", headers

  response = http.request request
  content = response.body
  p content.encoding
  p content.bytesize
  
  raise unless response["Content-Encoding"] == "gzip"
  content = Zlib::GzipReader.new(StringIO.new(content)).read
  p content.bytesize
end
```

```
$ ruby --vm.Xmx1g dl_versions.rb                
#<Encoding:ASCII-8BIT>
5957883
[ruby] WARNING OutOfMemoryError
gc.c:30:in `rb_tr_gc_guard': failed to allocate memory (NoMemoryError)
	from gc.c:30:in `rb_gc_force_recycle'
	from zlib.c:1099:in `zstream_run'
	from zlib.c:2697:in `gzfile_read_more'
	from zlib.c:2814:in `gzfile_read_all'
	from zlib.c:3813:in `rb_gzreader_read'
	from /home/eregon/.rubies/truffleruby-dev/lib/truffle/truffle/cext_ruby.rb:39:in `read'
	from dl_versions.rb:14:in `block in <main>'
	from /home/eregon/.rubies/truffleruby-dev/lib/mri/net/http.rb:933:in `start'
	from /home/eregon/.rubies/truffleruby-dev/lib/mri/net/http.rb:606:in `start'
	from dl_versions.rb:4:in `<main>'
```

**Comment by eregon** [comment]
@norswap found out that this results in computing many SubstringRope's of a large ConcatRope (of ConcatRopes)*.
We end up materializing all byte[] for each SubstringRope and so in the end duplicating each byte many times.
We're considering some ways to address that.

I've also found that Net::HTTP can decode gzip-encoded contents itself, and it does so in a streaming fashion which seems to consume much less memory than the naive approach of reading the whole String in memory, wrapping it in a StringIO and passing that to Zlib.
Actually, Net::HTTP does that automatically, but it's disabled by https://github.com/oracle/truffleruby/blob/69dc4f183d70fb40283fa6020c81b892a28b9f3f/lib/mri/bundler/compact_index_client/updater.rb#L50
That's trivial to change, just removing that line work so I'll merge that a short-term workaround for this issue.

**Comment by eregon** [comment]
This should be properly fixed now with ec62fae4f65ad57bf6de4d89ff70dd0597dbd058.
